internal/http/client.go

```diff
@@ -286,31 +287,63 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 		logger.Trace("request details", "headers", string(logHeaders))
 	}
 
-	resp, retryErr := c.client.Do(r)
-	if retryErr != nil {
-		return nil, retryErr
-	}
+	var shouldRetry bool
+	var body []byte
+	var resp *http.Response
+	var retryErr error
+	var readErr error
+	var errorValue ErrorResponse
 
-	defer resp.Body.Close()
+	for i := 0; ; i++ {
+		shouldRetry = false
 
-	if resp.StatusCode == http.StatusNotFound {
-		return nil, &nrErrors.NotFound{}
-	}
+		if i > 0 {
+			logger.Debug("retrying request", "method", req.method, "url", r.URL)
+		}
 
-	body, readErr := ioutil.ReadAll(resp.Body)
-	if readErr != nil {
-		return nil, readErr
-	}
+		resp, retryErr = c.client.Do(r)
+		if retryErr != nil {
+			return nil, retryErr
+		}
 
-	logHeaders, err = json.Marshal(resp.Header)
-	if err != nil {
-		return nil, err
-	}
+		defer resp.Body.Close()
```

Comment:
I think these won't be closed until all retries are finished. That's probably ok, but suboptimal. I usually put the body of my retry loop in a separate function so that I can use defer naturally.

Reply:
Agree. Updated 👍